File: matlab/src/solvers/MVA/solver_mva.m
function [Q,U,R,T,C,X,lG] = solver_mva(sn,options)
% [Q,U,R,T,C,X,LG] = SOLVER_MVA(SN,OPTIONS)

% Copyright (c) 2012-2026, Imperial College London
% All rights reserved.

if nargin < 2
    options = SolverMVA.defaultOptions;
end

[Lchain,STchain,Vchain,alpha,Nchain,~,refstatchain] = sn_get_demands_chain(sn);

nservers = sn.nservers;
sched = sn.sched;
M = sn.nstations;
K = sn.nchains;

% METHOD 'mva' IS THE DELIBERATE APPROXIMATION: SolverMVA.mvaDispatch already
% warns that the exact recursion is being run outside its hypotheses and
% promises an answer, so erroring here would contradict its own message. Only
% an implicit or 'exact' request is refused.
if ~sn_has_product_form(sn) && ~(isfield(options,'method') && strcmp(options.method,'mva'))
    line_error(mfilename, 'Unsupported exact MVA analysis, the model does not have a product form');
end

% Check for special LCFS + LCFS-PR 2-station network. The pairing, the station
% and server counts, the closed population and the self-loop rule are judged by
% SolverMVA.supportsLcfs, the predicate the report gates on, so a pair the
% report calls runnable is one that runs here.
lcfsStat = find(sched == SchedStrategy.LCFS);
if ~isempty(lcfsStat)
    [lcfsOk, lcfsReason] = SolverMVA.supportsLcfs(sn, 'mva');
    if ~lcfsOk
        line_error(mfilename, lcfsReason);
    end
    lcfsprStat = find(sched == SchedStrategy.LCFSPR);
    % Call specialized LCFS MVA solver
    [Q,U,R,T,C,X,lG] = solver_mva_lcfsqn(sn, options, lcfsStat, lcfsprStat);
    return;
end

infSET =[]; % set of infinite server stations
qSET =[]; % set of other product-form stations
for ist=1:M
    switch sched(ist)
        case SchedStrategy.EXT
            % no-op
        case SchedStrategy.INF
            infSET(1,end+1) = ist;
        case {SchedStrategy.PS, SchedStrategy.LCFSPR}
            qSET(1,end+1) = ist;
        case {SchedStrategy.FCFS, SchedStrategy.SIRO}
            qSET(end+1) = ist;
        otherwise
            line_error(mfilename, sprintf('Unsupported exact MVA analysis for %s scheduling.',SchedStrategy.toText(sched(ist))));
    end
end

Uchain = zeros(M,K); Tchain = zeros(M,K); C = zeros(1,K); Wchain = zeros(M,K); Qchain = zeros(M,K);

lambda = zeros(1,K);
ocl = find(isinf(Nchain));
if any(isinf(Nchain))
    for r=ocl % open classes
        lambda(r) = 1 ./ STchain(refstatchain(r),r);
        Qchain(refstatchain(r),r) = Inf;
    end
end
rset = setdiff(1:K,find(Nchain==0));

% Interlocked flow (Franks 1999, Eq. 4.7): a request cannot queue behind work
% that its own submission caused, so the arrival-instant queue drops the
% interlocked share of the other chains. SolverLN supplies the matrix.
IL = [];
if isfield(options,'config') && isfield(options.config,'interlock') && ~isempty(options.config.interlock)
    IL = sn_interlock_chain(sn, options.config.interlock);
end

% the interlocked recursion is a separate entry point: pfqn_mvams and the
% pfqn_mva family it dispatches to carry the standard arrival theorem only
if isempty(IL)
    [Xchain,Qpf,Uchain,~,lG] = pfqn_mvams(lambda,STchain(qSET,:).*Vchain(qSET,:),Nchain,STchain(infSET,:).*Vchain(infSET,:),ones(length(qSET),1),nservers(qSET));
else
    [Xchain,Qpf,Uchain,~,lG] = pfqn_mvams_ilock(lambda,STchain(qSET,:).*Vchain(qSET,:),Nchain,STchain(infSET,:).*Vchain(infSET,:),ones(length(qSET),1),nservers(qSET),IL);
end
Qchain(qSET,:) = Qpf;
Qchain(infSET,:) = repmat(Xchain,numel(infSET),1) .* STchain(infSET,:) .* Vchain(infSET,:);

ccl = find(isfinite(Nchain));
for r=rset
    for k=infSET(:)'
        Wchain(k,r) = STchain(k,r);
    end
    for k=qSET(:)'
        if isinf(nservers(k)) % infinite server
            Wchain(k,r) = STchain(k,r);
        else
            if Vchain(k,r) == 0 || Xchain(r) == 0
                Wchain(k,r) = 0;
            else
                Wchain(k,r) = Qchain(k,r) / (Xchain(r) * Vchain(k,r));
            end
        end
    end
end

for r=rset
    if sum(Wchain(:,r)) == 0
        Xchain(r) = 0;
    else
        if isinf(Nchain(r))
            C(r) = Vchain(:,r)'*Wchain(:,r);
            % X(r) remains constant
        elseif Nchain(r)==0
            Xchain(r) = 0;
            C(r) = 0;
        else
            C(r) = Vchain(:,r)'*Wchain(:,r);
            Xchain(r) = Nchain(r) / C(r);
        end
    end
    
    for k=1:M
        Qchain(k,r) = Xchain(r) * Vchain(k,r) * Wchain(k,r);
        Tchain(k,r) = Xchain(r) * Vchain(k,r);
    end
end

for k=1:M
    for r=rset
        if isinf(nservers(k)) % infinite server
            Uchain(k,r) = Vchain(k,r)*STchain(k,r)*Xchain(r);
        else
            Uchain(k,r) = Vchain(k,r)*STchain(k,r)*Xchain(r)/nservers(k);
        end
    end
end

for k=1:M
    for r=1:K
        if Vchain(k,r)*STchain(k,r) > options.tol
            switch sched(k)
                case {SchedStrategy.FCFS,SchedStrategy.PS}
                    if sum(Uchain(k,:))>1+options.tol
                        Uchain(k,r) = min(1,sum(Uchain(k,:))) * Vchain(k,r)*STchain(k,r)*Xchain(r) / ((Vchain(k,:).*STchain(k,:))*Xchain(:));
                    end
            end
        end
    end
end

%Vsink = cellsum(sn.nodevisits);
%Vsink = Vsink(find(sn.nodetype==NodeType.Sink),:);
%for r=find(isinf(Nchain)) % open classes
    %Xchain(r) = Vsink(r) ./ STchain(refstatchain(r),r);
%end

Rchain = Qchain./Tchain;
Xchain(~isfinite(Xchain))=0;
Uchain(~isfinite(Uchain))=0;
Qchain(~isfinite(Qchain))=0;
Rchain(~isfinite(Rchain))=0;

Xchain(Nchain==0)=0;
Uchain(:,Nchain==0)=0;
Qchain(:,Nchain==0)=0;
Rchain(:,Nchain==0)=0;
Tchain(:,Nchain==0)=0;
Wchain(:,Nchain==0)=0;

[Q,U,R,T,C,X] = sn_deaggregate_chain_results(sn, Lchain, [], STchain, Vchain, alpha, [], [], Rchain, Tchain, [], Xchain);
end

File: jar/src/main/kotlin/jline/solvers/mva/handlers/Solver_mva.kt
/*
 * Copyright (c) 2012-2026, QORE Lab, Imperial College London
 * All rights reserved.
 */
package jline.solvers.mva.handlers

import jline.GlobalConstants
import jline.GlobalConstants.Inf
import jline.api.pfqn.mva.pfqn_mvams
import jline.api.sn.snDeaggregateChainResults
import jline.api.sn.snGetDemandsChain
import jline.api.sn.snHasProductForm
import jline.lang.NetworkStruct
import jline.lang.constant.NodeType
import jline.lang.constant.SchedStrategy
import jline.solvers.SolverOptions
import jline.solvers.mva.MVAResult
import jline.util.Maths
import jline.util.Utils
import jline.util.matrix.Matrix
import java.lang.Double
import kotlin.Int
import kotlin.RuntimeException


/**
 * Handler for the solver_mva function.
 */
fun solver_mva(sn: NetworkStruct, options: SolverOptions): MVAResult {
    val ret = snGetDemandsChain(sn)
    val Lchain = ret.Dchain
    val STchain = ret.STchain
    val Vchain = ret.Vchain
    val alpha = ret.alpha
    val Nchain = ret.Nchain
    val refstatchain = ret.refstatchain
    val nservers = sn.nservers
    // We're using schedStrategies instead of schedid, as schedid is not available in NetworkStruct
    val M = sn.nstations
    val K = sn.nchains

    // Check for LCFS scheduling - not supported in this version
    for (i in 0..<M) {
        when (sn.sched.get(sn.stations.get(i))) {
            SchedStrategy.LCFS, SchedStrategy.LCFSPR ->
                throw RuntimeException("LCFS queueing networks are not supported in this version.")
            else -> {}
        }
    }

    // For non-LCFS models, check product-form requirement
    val infSET = ArrayList<Int?>() // set of infinite server stations
    val qSET = ArrayList<Int?>() // set of other product-form stations
    if (!snHasProductForm(sn)){
        throw RuntimeException("Unsupported exact MVA analysis, the model does not have a product form")
    }

    for (i in 0..<M) {
        when (sn.sched.get(sn.stations.get(i))) {
            SchedStrategy.EXT -> {}
            SchedStrategy.INF -> infSET.add(i)
            SchedStrategy.PS, SchedStrategy.LCFSPR -> qSET.add(i)
            SchedStrategy.FCFS, SchedStrategy.SIRO -> qSET.add(i)
            else -> throw RuntimeException("Unsupported exact MVA analysis for " + SchedStrategy.toText(sn.sched.get(sn.stations.get(
                i))) + " scheduling")
        }
    }
    val Uchain = Matrix(M, K)
    val Tchain = Matrix(M, K)
    val C = Matrix(1, K)
    val Wchain = Matrix(M, K)
    val Qchain = Matrix(M, K)
    val lambda = Matrix(1, K)

    val ocl = ArrayList<Int?>()
    for (i in 0..<Nchain.numCols) {
        if (Utils.isInf(Nchain.get(i))) {
            ocl.add(i)
        }
    }
    for (r in ocl) { // open classes
        lambda.set(0, r!!, 1.0 / STchain.get(refstatchain.get(r).toInt(), r))
        Qchain.set(refstatchain.get(r).toInt(), r, Inf)
    }
    val rset = ArrayList<Int?>()
    for (i in 0..<K) {
        if (Nchain.get(i) != 0.0) {
            rset.add(i)
        }
    }
    val Lp = Matrix(qSET.size, STchain.numCols)
    for (i in qSET.indices) {
        for (j in 0..<Lp.numCols) {
            Lp.set(i, j, STchain.get(qSET.get(i)!!, j) * Vchain.get(qSET.get(i)!!, j))
        }
    }
    val Zp = Matrix(infSET.size, STchain.numCols)
    for (i in infSET.indices) {
        for (j in 0..<Zp.numCols) {
            Zp.set(i, j, STchain.get(infSET.get(i)!!, j) * Vchain.get(infSET.get(i)!!, j))
        }
    }
    val nserversp = Matrix(qSET.size, 1)
    for (i in qSET.indices) {
        nserversp.set(i, nservers.get(qSET.get(i)!!))
    }
    val ret1 = pfqn_mvams(lambda, Lp, Nchain, Zp, Matrix.ones(qSET.size, 1), nserversp)
    val Xchain = ret1.X
    val Qpf = ret1.Q/* Uchain = ret1.UN; -> not needed */
    val lG = ret1.lGN
    for (i in qSET.indices) {
        for (j in 0..<Qchain.numCols) {
            Qchain.set(qSET.get(i)!!, j, Qpf.get(i, j))
        }
    }

    val Q2 = Matrix(infSET.size, Qchain.numCols)
    val Xchainrep = Xchain.repmat(infSET.size, 1)
    for (i in infSET.indices) {
        for (j in 0..<Q2.numCols) {
            Q2.set(i, j, Xchainrep.get(i, j) * STchain.get(infSET.get(i)!!, j) * Vchain.get(infSET.get(i)!!, j))
        }
    }

    for (i in infSET.indices) {
        for (j in 0..<Qchain.numCols) {
            Qchain.set(infSET.get(i)!!, j, Q2.get(i, j))
        }
    }

    val ccl = ArrayList<Int?>()
    for (i in 0..<Nchain.numCols) {
        if (Double.isFinite(Nchain.get(i))) {
            ccl.add(i)
        }
    }
    for (r in rset) {
        for (k in infSET) {
            Wchain.set(k!!, r!!, STchain.get(k, r))
        }
        for (k in qSET) {
            if (Utils.isInf(nservers.get(k!!))) { // Infinite server
                Wchain.set(k, r!!, STchain.get(k, r))
            } else {
                if (Vchain.get(k, r!!) == 0.0 || Xchain.get(r) == 0.0) {
                    Wchain.set(k, r, 0)
                } else {
                    Wchain.set(k, r, Qchain.get(k, r) / (Xchain.get(r) * Vchain.get(k, r)))
                }
            }
        }
    }
    for (r in rset) {
        if (Matrix.extractColumn(Wchain, r!!, null).elementSum() == 0.0) {
            Xchain.set(r, 0.0)
        } else {
            if (Utils.isInf(Nchain.get(r))) {
                val vt = Matrix.extractColumn(Vchain, r, null)
                val wt = Matrix.extractColumn(Wchain, r, null)
                C.set(r, vt.transpose().mult(wt).get(0))
            } else if (Nchain.get(r) == 0.0) {
                Xchain.set(r, 0.0)
                C.set(r, 0.0)
            } else {
                val vt = Matrix.extractColumn(Vchain, r, null)
                val wt = Matrix.extractColumn(Wchain, r, null)
                C.set(r, vt.transpose().mult(wt).get(0))
                Xchain.set(r, Nchain.get(r) / C.get(r))
            }
        }

        for (k in 0..<M) {
            Qchain.set(k, r, Xchain.get(r) * Vchain.get(k, r) * Wchain.get(k, r))
            Tchain.set(k, r, Xchain.get(r) * Vchain.get(k, r))
        }
    }
    for (k in 0..<M) {
        for (r in rset) {
            if (Utils.isInf(nservers.get(k))) { // infinite server
                Uchain.set(k, r!!, Vchain.get(k, r) * STchain.get(k, r) * Xchain.get(r))
            } else {
                Uchain.set(k, r!!, Vchain.get(k, r) * STchain.get(k, r) * Xchain.get(r) / nservers.get(k))
            }
        }
    }
    for (k in 0..<M) {
        for (r in 0..<K) {
            if (Vchain.get(k, r) * STchain.get(k, r) > options.tol) {
                when (sn.sched.get(sn.stations.get(k))) {
                    SchedStrategy.FCFS, SchedStrategy.PS -> {
                        val Urow = Matrix.extractRows(Uchain, k, k + 1, null)
                        val UrowSum = Urow.elementSum()
                        if (UrowSum > 1 + options.tol) {
                            val Vrow = Matrix.extractRows(Vchain, k, k + 1, null)
                            val STrow = Matrix.extractRows(STchain, k, k + 1, null)
                            Uchain.set(k,
                                r,
                                Maths.min(1.0, UrowSum) * Vchain.get(k, r) * STchain.get(k,
                                    r) * Xchain.get(r) / Vrow.elementMult(STrow, null).mult(Xchain.columnMajorOrder())
                                    .get(0))
                        }
                    }
                    else -> {}
                }
            }
        }
    }
    var Vs: Matrix? = null
    for (key in sn.nodevisits.keys) {
        if (Vs == null) {
            Vs = sn.nodevisits.get(key)
        } else {
            Vs = Vs.add(1.0, sn.nodevisits.get(key))
        }
    }
    val sinks = ArrayList<Int?>()
    for (i in sn.nodetype.indices) {
        if (sn.nodetype.get(i) == NodeType.Sink) {
            sinks.add(i)
        }
    }
//    val Vsink = Matrix(sinks.size, Vs!!.getNumCols())
//    for (i in sinks.indices) {
//        for (j in 0..<Vsink.getNumCols()) {
//            Vsink.set(i, j, Vs.get(sinks.get(i)!!, j))
//        }
//    }
//    for (r in 0..<Nchain.getNumCols()) {
//        if (Utils.isInf(Nchain.get(r))) { // open classes
//            Xchain.set(r, Vsink.get(r) / STchain.get(refstatchain.get(r).toInt(), r))
//        }
//    }

    val Rchain = Matrix(Qchain.numRows, Qchain.numCols)
    for (i in 0..<Rchain.numRows) {
        for (j in 0..<Rchain.numCols) {
            if (Qchain.get(i, j) < GlobalConstants.Zero) {
                Rchain.set(i, j, 0.0)
            } else {
                Rchain.set(i, j, Qchain.get(i, j) / Tchain.get(i, j))
            }
        }
    }

    for (i in 0..<Xchain.numRows) {
        for (j in 0..<Xchain.numCols) {
            if (!Double.isFinite(Xchain.get(i, j))) Xchain.set(i, j, 0)
        }
    }

    for (i in 0..<Uchain.numRows) {
        for (j in 0..<Uchain.numCols) {
            if (!Double.isFinite(Uchain.get(i, j))) Uchain.set(i, j, 0)
        }
    }

    for (i in 0..<Qchain.numRows) {
        for (j in 0..<Qchain.numCols) {
            if (!Double.isFinite(Qchain.get(i, j))) Qchain.set(i, j, 0)
        }
    }

    for (i in 0..<Rchain.numRows) {
        for (j in 0..<Rchain.numCols) {
            if (!Double.isFinite(Rchain.get(i, j))) Rchain.set(i, j, 0)
        }
    }

    val Nzero = ArrayList<Int?>()
    for (i in 0..<Nchain.numCols) {
        if (Nchain.get(i) == 0.0) Nzero.add(i)
    }
    for (j in Nzero) {
        Xchain.set(j!!, 0.0)
        for (i in 0..<Uchain.numRows) {
            Uchain.set(i, j, 0.0)
        }
        for (i in 0..<Qchain.numRows) {
            Qchain.set(i, j, 0.0)
        }
        for (i in 0..<Rchain.numRows) {
            Rchain.set(i, j, 0.0)
        }
        for (i in 0..<Tchain.numRows) {
            Tchain.set(i, j, 0.0)
        }
        for (i in 0..<Wchain.numRows) {
            Wchain.set(i, j, 0.0)
        }
    }

    val ret2 =
        snDeaggregateChainResults(sn, Lchain, null, STchain, Vchain, alpha, null, null, Rchain, Tchain, null, Xchain)

    val res = MVAResult()
    res.QN = ret2.Q
    res.UN = ret2.U
    res.RN = ret2.R
    res.TN = ret2.T
    res.CN = ret2.C
    res.XN = ret2.X
    res.logNormConstAggr = lG
    return res
}

